File: followers/app/rabbitmq_wrapper.py
import asyncio
import aio_pika
import aio_pika.abc
from typing import Annotated
from app.utils.message_handler import message_handler


class RabbitMQWrapper:
    __connection: aio_pika.abc.AbstractConnection = None
    __channel: aio_pika.abc.AbstractChannel
    
    def get_client(self):
        if(self.__connection is None):
            raise Exception("Cannot get client before connection")
        else:
            return self.__connection

    async def connect(self):
        if self.__connection is not None:
            return self.__connection
        for i in range(5):
            try:
                self.__connection = await aio_pika.connect_robust("amqp://jaysurya00:muskansinghvi@myrabbit:5672")
                print("Connected to RabbitMQ")
                self.__channel= await self.__connection.channel()
                return self.__connection
            except Exception as err:
                print(f"Connection attempt to RabbitMQ failed. Retries{i}")
                await asyncio.sleep(15)
        raise Exception(
            "Connection attempt to RabbitMQ failed after several reattempts"
        )

    async def consume(self, queue_name: str):
        queue = await self.__channel.declare_queue(queue_name, durable=True)
        await queue.consume(message_handler)
        print(f"Started consuming from queue: {queue_name}")
        
    async def close(self):
        if self.__connection:
            await self.__connection.close()


rabbitmqWrapper = RabbitMQWrapper()

File: feeds-stream-processor/src/rabbitmqWrapper.ts
import client, { Connection, Channel, ConsumeMessage, Message } from 'amqplib';
import msgHandler from './utils/handlerPostMsg';

class RabbitMQWrapper {
    private _connection?: Connection;

    getClient() {
        if (!this._connection) {
            throw new Error("Cannot access RabbitMQ client before connecting")
        }
        return this._connection;
    }

    async connect() {
        if(this._connection) return this._connection
        for (let i = 1; i < 6; i++) {
            try {
                this._connection = await client.connect('amqp://jaysurya00:muskansinghvi@myrabbit:5672');
                console.log("Connected to RabbitMQ")
                return this._connection;
            }
            catch (err) {
                console.log("Connection attempt failed to RabbitMQ. Retrying...");
                await new Promise((resolve, reject)=>{
                    setTimeout(resolve, 5000)
                })
            }
        }
        throw new Error("Failed to connect RabbitMQ after several reattempts");
    }

    async consume(queue: string) {
        if (!this._connection) {
            throw new Error("Cannot create channel before connecting");
        }
        try {
            const channel: Channel = await this._connection.createChannel();
            channel.assertQueue(queue);
            channel.consume(queue,async (msg)=>{
                if (msg !== null) {
                    msgHandler(msg);
                    channel.ack(msg);
                  } else {
                    console.log('Consumer cancelled by server');
                  }
            });
            console.log(`Started consuming from the queue ${queue}`);
        }
        catch (err) {
            throw new Error("Publishing event failed " + err);
        }
    }
}

const rabbitmqWrapper = new RabbitMQWrapper();
export default rabbitmqWrapper;